### PR TITLE
sql: switch to settings for stats collection

### PR DESCRIPTION
**Pull request by dt**

### CODE REVIEW

File: pkg/sql/app_stats.go [hunk 1]
```diff
@@ -230,20 +241,24 @@ func dumpStmtStats(ctx context.Context, appName string, stats map[string]*stmtSt
...
+	"frequency at which per-app and per-statement statistics are reset",
+	1*time.Hour,
 )
 
 // startResetWorker ensures that the data is removed from memory
 // periodically, so as to avoid memory blow-ups.
 func (s *sqlStats) startResetWorker(stopper *stop.Stopper) {
 	ctx := log.WithLogTag(context.Background(), "sql-stats", nil)
 	stopper.RunWorker(ctx, func(ctx context.Context) {
-		ticker := time.NewTicker(StmtStatsResetFrequency)
```
Comment (tamird) [inline pkg/sql/app_stats.go]: do we need to lint for this?

Comment (dt) [inline pkg/sql/app_stats.go]: because of the issues with stdlib timer in general? or just for the fixed-rate-ness of Ticker?

Comment (tamird) [inline pkg/sql/app_stats.go]: The former. I'm wondering if we should (and why we haven't) ban time.Timer, time.Ticker, and time.After.

Comment (dt) [inline pkg/sql/app_stats.go]: Maybe? I don't have strong opinions either way.

**Comment by knz** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/app_stats.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-KiKLcxaX4lG6TgIMeEB:-KiKLcxaX4lG6TgIMeEC:b-ts4e5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd8ded8909a7520aab94c97999174ce473c821d/pkg/sql/app_stats.go#L57)):*
> ```Go
> // consumed by a SQL statement before it is collected for statistics reporting.
> var SQLStatsCollectionLatencyThreshold = settings.RegisterDurationSetting(
> 	"sql.metrics.stmtstats.threshold",
> ```

Simplify: use only the threshold and default it to 1ns.

---

*[pkg/sql/app_stats.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-KiKLsHDVgxJnNLqo-cd:-KiKLsHEc4tahvVcdXKZ:b-htm82e) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd8ded8909a7520aab94c97999174ce473c821d/pkg/sql/app_stats.go#L245)):*
> ```Go
> // unlimited memory growth.
> var StmtStatsResetFrequency = settings.RegisterDurationSetting(
> 	"sql.metrics.stmtstats.interval",
> ```

`stmtstats.reset.interval`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252)*

File: pkg/sql/app_stats.go (hunk 1 above)
Comment (bdarnell) [inline pkg/sql/app_stats.go]: We should probably ban time.Timer, and in fact I thought we already did. `time.After` is a little inefficient, but I don't think it's worth banning and replacing with a cancelable alternative. What's wrong with `time.Ticker` aside from the unchangeable rate?

File: pkg/sql/app_stats.go [hunk 2]
```diff
@@ -46,14 +47,16 @@ type stmtStats struct {
 
 // StmtStatsEnable determines whether to collect per-statement
 // statistics.
-var StmtStatsEnable = envutil.EnvOrDefaultBool(
-	"COCKROACH_SQL_STMT_STATS_ENABLE", true,
+var StmtStatsEnable = settings.RegisterBoolSetting(
+	"sql.metrics.stmtstats.enabled", "collect per-statement query statistics ", true,
```
Comment (bdarnell) [inline pkg/sql/app_stats.go]: I'd spell `statement` in full, and put a separator between it and `stats`. Putting `metrics` and `stats` in the same name seems redundant.

Comment (dt) [inline pkg/sql/app_stats.go]: maybe something like `sql.metrics.record_per_statement_latencies` ?

Comment (knz) [inline pkg/sql/app_stats.go]: There is more data collected than just latencies.

Comment (dt) [inline pkg/sql/app_stats.go]: any name suggestions?

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/app_stats.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-KiGdDXO7DQrFBO5Eak2-r1-51:-KiVdswQH1ZhIMXMe_eW:b4uyoa4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd8ded8909a7520aab94c97999174ce473c821d/pkg/sql/app_stats.go#L51)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

any name suggestions?
</blockquote></details>

I'll take `sql.metrics.recordStatementDetails` or `sql.metrics.record-statements-details`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/app_stats.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-KiKLsHDVgxJnNLqo-cd:-KiVezIo25NfTMR5l5oU:b-xundm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd8ded8909a7520aab94c97999174ce473c821d/pkg/sql/app_stats.go#L245)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`stmtstats.reset.interval`
</blockquote></details>

`sql.metrics.statement-details.reset-interval`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252)*

**Comment by dt** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/app_stats.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-KiGdDXO7DQrFBO5Eak2-r1-51:-KiVeGKujLY2V1dwRPB0:b20u9t8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd8ded8909a7520aab94c97999174ce473c821d/pkg/sql/app_stats.go#L51)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'll take `sql.metrics.recordStatementDetails` or `sql.metrics.record-statements-details`
</blockquote></details>

currently we lower-case during lookup for SET so currently all settings must be defined as lower only.

We could undo that and have exact-case lookup, or add a second map case-folded, but I guess we should decide over on your naming PR what we want for word separators and then go with that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/app_stats.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-KiGdDXO7DQrFBO5Eak2-r1-51:-KiVfXDbbBVkFTlpEMow:b-5lvkkl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd8ded8909a7520aab94c97999174ce473c821d/pkg/sql/app_stats.go#L51)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

currently we lower-case during lookup for SET so currently all settings must be defined as lower only.

We could undo that and have exact-case lookup, or add a second map case-folded, but I guess we should decide over on your naming PR what we want for word separators and then go with that.
</blockquote></details>


let's keep the case insensitivity, I like it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252)*

**Comment by dt** [comment]
Removed from this change -- this loop probably gets absorbed in the diagnostic reporting loop anyway.

---

Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/app_stats.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-KiGdDXO7DQrFBO5Eak2-r1-51:-Ki_ub8b2WQC37f9SWKI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd8ded8909a7520aab94c97999174ce473c821d/pkg/sql/app_stats.go#L51)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

let's keep the case insensitivity, I like it.
</blockquote></details>

Done.

---

*[pkg/sql/app_stats.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-KiKLcxaX4lG6TgIMeEB:-Ki_uZd0HojVV9ahN1rC:b-kj6i6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd8ded8909a7520aab94c97999174ce473c821d/pkg/sql/app_stats.go#L57)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Simplify: use only the threshold and default it to 1ns.
</blockquote></details>

I think I like the explicit knobs.

---

*[pkg/sql/app_stats.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-KiKLsHDVgxJnNLqo-cd:-Ki_uXW2PuUKV2NpIS0R:b23nd1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd8ded8909a7520aab94c97999174ce473c821d/pkg/sql/app_stats.go#L245)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`sql.metrics.statement-details.reset-interval`?
</blockquote></details>

Removed this from the change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-:-Ki_ucJq-gksTIAnKFAR:bqoertv)*

**Comment by knz** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/app_stats.go, line 177 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-Kia-yGIqtN3jQQvvlrs:-Kia-yGIqtN3jQQvvlrt:bpb8qlk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3a3058725b35cb6fd49c51fe752854f63234c7/pkg/sql/app_stats.go#L177)):*
> ```Go
> 
> var dumpStmtStatsToLogBeforeReset = settings.RegisterBoolSetting(
> 	"sql.metrics.statement_details.dumptologs",
> ```

`dump_to_logs` then I suppose

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252)*

**Comment by dt** [comment]
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/app_stats.go, line 177 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-Kia-yGIqtN3jQQvvlrs:-Kia0GUaPX_exDBkTC66:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3a3058725b35cb6fd49c51fe752854f63234c7/pkg/sql/app_stats.go#L177)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`dump_to_logs` then I suppose
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252)*

**Comment by knz** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-:-Kia0U2sfZgj4a9KPCRJ:bnfp4nl)*

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/app_stats.go, line 177 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-Kia1v7B_VNh-SnfEo0S:-Kia1v7B_VNh-SnfEo0T:bpv12mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e92f686af24c04d4bd628031df50978e895a498c/pkg/sql/app_stats.go#L177)):*
> ```Go
> 
> var dumpStmtStatsToLogBeforeReset = settings.RegisterBoolSetting(
> 	"sql.metrics.statement_details.dump_to_logs",
> ```

`statement` is singular here but plural in the other two. I prefer the singular.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252)*

**Comment by dt** [comment]
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/app_stats.go, line 177 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252#-Kia1v7B_VNh-SnfEo0S:-Kia2ius4EARhS7bJwvU:b-kfyliq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e92f686af24c04d4bd628031df50978e895a498c/pkg/sql/app_stats.go#L177)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`statement` is singular here but plural in the other two. I prefer the singular.
</blockquote></details>

whoops, nice catch, thanks.
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15252)*

**Pull request merged**